### ISSUE TITLE
AccessViolationException in zlib on CLI Windows build

### ISSUE THREAD

**Issue by eerhardt** [dotnet/corefx#9265]
## Repro steps:

- step 1: get a CLI repo
- step 1.a: add a remote for github.com/dotnet/cli
- step 2: git checkout -b preview2 upstream/rel/1.0.0-preview2
  - In case the cli repo moves forward, and somehow that causes the AV to no longer occur:  use this sha1: 3c9f8dd99be5918a504f34260f4d6016a086ea99
- step 3: build.cmd -Targets Prepare,Compile
  - that should succeed
- step 4: build.cmd -Targets Prepare,InitPackage,GenerateVersionBadge,GenerateZip,GenerateMsis,GenerateNugetPackages
  - that will hit the AV

## Notes

I’ve uploaded a mini dump with heap:

`\\scratch2\scratch\eerhardt\dotnet.dmp`

I was able to find clrcompression symbols on the MyGet symbol server:  https://dotnet.myget.org/F/dotnet-core/symbols/

Unfortunately, talking with @dagood we are not able to locate the coreclr.pdb file matching the build from yesterday’s CoreCLR build:  1.0.24207.2.

I still have the process broken on my machine with an attached debugger, here is the callstack:

```
>             clrcompression.dll!fizzle_matches(internal_state * s, match * current, match * next) Line 121       C
               clrcompression.dll!deflate_medium(internal_state * s, int flush) Line 265               C
               clrcompression.dll!deflate(z_stream_s * strm, int flush) Line 982 C
               00007ff7c22e8403()         Unknown
               0000020f85a48610()        Unknown
               00000016a837cb40()       Unknown
               0000020f876fd5b8()        Unknown
               0000020f876ff710()         Unknown
               0000be2449de0620()      Unknown
               coreclr.dll!00007ff8220276b0()   Unknown
```

line 121 is:

```C
static void fizzle_matches(deflate_state *s, struct match *current, struct match *next)
{
    IPos limit;
    unsigned char *match, *orig;
    int changed = 0;
    struct match c,n;
    /* step zero: sanity checks */

    if (current->match_length <= 1)
            return;

    match = s->window - current->match_length + 1 + next->match_start ;
    orig  = s->window - current->match_length + 1 + next->strstart ;
    
    /* quick exit check.. if this fails then don't bother with anything else */
    if (zlikely(*match != *orig))   <<<<< HERE
        return;
```

/cc @ianhays @stephentoub @Petermarcu

**Comment by stephentoub** [comment]
cc: @bjjones

**Comment by ianhays** [comment]
I got this reproducing on a fresh CLI repo install. 

I figure there are two potential places where this could be failing as a result of clrcompression related issues:
- The Windows Open native build is doing something wrong when building the dll.
- Something in zlib is producing an error.

To test the open build I tried building vanilla zlib 1.2.8 using the open native build and placing the result dll in my CLI repo. I could no longer reproduce the issue, so the native build is likely fine.

That means we've got an issue in the zlib sources themselves. It only fails for me about one fifth of the time, but it's relatively consistent on that interval. The failure isn't happening in the tests, so it's the result of something that the CLI repo is doing that we don't cover. Looking into it more.

**Comment by ianhays** [comment]
I can't reproduce this locally any more. My past ~20 builds on a fresh CLI repo have passed.

**Comment by bjjones** [comment]
I haven't been able to reproduce yet (due to lack of experience building cli), however I found this seemingly related commit in a branch on the Zlib-Intel repo. It looks like the fix was overlooked and never got merged into master (which got merged to corefx).

https://github.com/jtkukunas/zlib/commit/e6a97efba0394dee35b1460e50978174ebb12126

I'm working to confirm this now.

**Comment by eerhardt** [comment]
@bjjones - if you need help setting up the CLI repo, let me know.  Machine instructions are:  https://github.com/dotnet/cli/blob/rel/1.0.0/Documentation/developer-guide.md#for-windows

Basically, you need:
- VS 2015 installed
- Cmake installed on and on your path
- Git installed and on your path.

**Comment by ianhays** [comment]
Thanks @bjjones, that looks like it might be our issue.

Similar to that, what about this other unmerged commit: https://github.com/jtkukunas/zlib/commit/fa19b0b10631646c30a7aa6f2ecf4b10d363e782 ? Are there other fixes in the zlib-intel repo we should mirror over to CoreFX?

**Comment by bjjones** [comment]
@eerhardt Thanks for posting that. I was on a machine with only VS13 :)

@ianhays, that commit pertains only to Intel's own compression strategy - deflate_quick() - for compressionLevel=1. When I initially merged in Zlib-Intel, I disabled deflate_quick() in favor of the conventional deflate() (same as vanilla Zlib) due to some compression ratio concerns. So corefx isn't using that code, however it couldn't hurt to make the change anyway for parity's sake. It looks like the other issue-related branches have been merged in already, but I'll ask around to be sure.

I have been able to build CLI ~20 times and haven't yet seen the AV, but I'll have a PR up soon to merge the mentioned Zlib-Intel fixes.

**Comment by ianhays** [comment]
Thanks Brandon. I did some more testing and applying the changes from jtkukunas/zlib@e6a97ef seems to resolve the problem.

I wrote a little script to run dotnet pack in an infinite while loop on the same nupkgs that are built by the CLI and tested it on three clrcompression versions:
- clrcompression x64 from the package
- clrcompression x64 locally built
- clrcompression x64 locally built with jtkukunas/zlib@fa19b0b applied

The first two failed intermittently and very rarely - about 1% of the time for me.
The clrcompression with the change did not fail in ~5000 dotnet pack calls. I'm going to keep running it to be sure, but it looks good.

**Comment by joshfree** [comment]
we're making a conservative fix to unblock CLI and downstream teams tonight;  we can continue to work and review the zlib-intel fixes in parallel.  But for now we'll revert the change from 2 days earlier which got us into the state where CLI builds are basically 100% blocked

**Comment by ianhays** [comment]
Closing this issue - the failures were resolved in the Release branch via #9289
